protocols/raft/src/test/java/io/atomix/protocols/raft/RaftTest.java

```diff
@@ -967,8 +964,41 @@ public void testThreeNodesAndRestartFollower() throws Throwable {
 
     for (int i = 0; i < 1_000; i++) {
       primitive.sendEvent(true).thenRun(this::resume);
-      await(30000, 2);
     }
+    await(30000, 2 * 1000);
+
+    // when
+    LoggerFactory.getLogger(RaftTest.class).error("====\nRestart!\n====");
+    members.removeIf(r -> r.memberId().equals(memberId));
+    createServer(memberId).bootstrap(members.stream().map(RaftMember::memberId).collect(Collectors.toList())).thenRun(this::resume);
+
+    // then
+    await(30000);
+  }
+
+  /**
+   * Tests submitting linearizable events.
+   */
+  @Test
+  public void testThreeNodesAndRestartLeader() throws Throwable {
+    // given
+    List<RaftServer> servers = createServers(3);
+
+    RaftClient client = createClient();
+    TestPrimitive primitive = createPrimitive(client);
+    primitive.onEvent(event -> {
+      threadAssertNotNull(event);
+      resume();
+    });
+
+    RaftServer leader = servers.stream().filter(s -> s.getRole() == Role.LEADER).findFirst().get();
+    final MemberId memberId = new MemberId(leader.name());
+    leader.shutdown().get(10, TimeUnit.SECONDS);
+
+    for (int i = 0; i < 1_000; i++) {
+      primitive.sendEvent(true).thenRun(this::resume);
```

Comment:
What does it test actually? I am wondering if it really has to write so many events.

Reply:
That shutdown and bootstrap (join) works even if there was a lot of new events in between. Most of the tests in atomix uses clean leave and rejoin, which will not happen in the real world :sweat_smile:

Reply:
I think I can also reduce the events a bit

Reply:
> That shutdown and bootstrap (join) works even if there was a lot of new events in between. Most of the tests in atomix uses clean leave and rejoin, which will not happen in the real world sweat_smile

:+1: Make sense.